src/flb_utils.c

```diff
@@ -732,20 +743,102 @@ int flb_utils_write_str(char *buf, int *off, size_t size,
             i += (hex_bytes - 1);
         }
         else if (c > 0xFFFF) {
```

Comment:
@matthewfala , I am sorry I only got time to take a closer look at this today. Would you please hold off on merging this change till we can talk on Monday?
I have a few concerns and questions:
1. I still think this change addresses the symptom and not the root cause of the bug. There is something else that is corrupting the utf sequence before we reach this point.
2. I don’t fully follow what `c` here is supposed to represent but comparing it with 0xFFFF seems like we are trying to handle UTF8 character sequences longer than 4 bytes. Perhaps you could explain it to me.
3. The code doesn’t validate for a trailing zero after leading 1s in the first byte.
4. Why are truncated sequences handled differently from invalid sequences? The behavior should be consistent for any invalid utf8 sequence.
5. This function has become very long and is hard to follow. I can share some thoughts on refactoring when we talk.
6. Do all log destinations handle the private use areas correctly? I preferred the replacement character approach.

Note: I didn’t actually run your code yet or write tests so I could be wrong about some of those points. Correct me if that is the case.

Reply:
2. Yes, that is how the cast works today and by that very definition, the `else if (c >= 0x80 && c <= 0xFFFF) {` case will never get executed.